src/transport/tls.rs

```diff
@@ -1,111 +1,296 @@
-use std::net::SocketAddr;
-
-use super::{AddrMaybeCached, SocketOpts, TcpTransport, Transport};
-use crate::config::{TlsConfig, TransportConfig};
-use crate::helper::host_port_pair;
-use anyhow::{anyhow, Context, Result};
-use async_trait::async_trait;
-use std::fs;
-use tokio::net::{TcpListener, TcpStream, ToSocketAddrs};
-use tokio_native_tls::native_tls::{self, Certificate, Identity};
-use tokio_native_tls::{TlsAcceptor, TlsConnector, TlsStream};
-
-#[derive(Debug)]
-pub struct TlsTransport {
-    tcp: TcpTransport,
-    config: TlsConfig,
-    connector: Option<TlsConnector>,
-    tls_acceptor: Option<TlsAcceptor>,
-}
+#[cfg(all(feature = "native-tls", feature = "rustls"))]
+compile_error!("Only one of `native-tls` and `rustls` can be enabled");
 
-#[async_trait]
-impl Transport for TlsTransport {
-    type Acceptor = TcpListener;
-    type RawStream = TcpStream;
-    type Stream = TlsStream<TcpStream>;
-
-    fn new(config: &TransportConfig) -> Result<Self> {
-        let tcp = TcpTransport::new(config)?;
-        let config = config
-            .tls
-            .as_ref()
-            .ok_or_else(|| anyhow!("Missing tls config"))?;
-
-        let connector = match config.trusted_root.as_ref() {
-            Some(path) => {
-                let s = fs::read_to_string(path)
-                    .with_context(|| "Failed to read the `tls.trusted_root`")?;
-                let cert = Certificate::from_pem(s.as_bytes())
-                    .with_context(|| "Failed to read certificate from `tls.trusted_root`")?;
-                let connector = native_tls::TlsConnector::builder()
-                    .add_root_certificate(cert)
-                    .build()?;
-                Some(TlsConnector::from(connector))
-            }
-            None => {
-                // if no trusted_root is specified, allow TlsConnector to use system default
-                let connector = native_tls::TlsConnector::builder().build()?;
-                Some(TlsConnector::from(connector))
-            }
-        };
+#[cfg(feature = "native-tls")]
```

Comment:
is it possible to split two tls implementation into two files? That will definitely look much better. For example, `native_tls.rs` and `rustls.rs`, and we import and alias one of them as the final tls transport in `tls.rs`. Even in subdirectory `tls`, if that's more convenient.

Reply:
Done